pkg/indexer/atlasprojectsecrets.go

```diff
@@ -0,0 +1,72 @@
+package indexer
+
+import (
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1/common"
+)
+
+const (
+	AtlasProjectBySecretsIndex = "atlasproject.spec.secrets"
+)
+
+type AtlasProjectByConnectionSecretIndexer struct {
+	logger *zap.SugaredLogger
+}
+
+func NewAtlasProjectByConnectionSecretIndexer(logger *zap.Logger) *AtlasProjectByConnectionSecretIndexer {
+	return &AtlasProjectByConnectionSecretIndexer{
+		logger: logger.Named(AtlasProjectBySecretsIndex).Sugar(),
+	}
+}
+
+func (AtlasProjectByConnectionSecretIndexer) Object() client.Object {
+	return &akov2.AtlasProject{}
+}
+
+func (*AtlasProjectByConnectionSecretIndexer) Name() string {
+	return AtlasProjectBySecretsIndex
+}
+
+func (a *AtlasProjectByConnectionSecretIndexer) Keys(object client.Object) []string {
+	project, ok := object.(*akov2.AtlasProject)
+	if !ok {
+		a.logger.Errorf("expected *akov2.AtlasProject but got %T", object)
+		return nil
+	}
+
+	result := sets.New[string]()
+	addIfNotEmpty := func(ref *common.ResourceRefNamespaced) {
+		if !ref.IsEmpty() {
+			result.Insert(ref.GetObject(project.Namespace).String())
+		}
+	}
```

Comment:
nitpick

Why is this not a regular private function?

```go
	func addIfNotEmpty(result sets.Set, ref *common.ResourceRefNamespaced) {
		if !ref.IsEmpty() {
			result.Insert(ref.GetObject(project.Namespace).String())
		}
	}
```

I do not see the reason for this to be a closure. Am I missing something?

Reply:
yes, this is enclosing over `result`

Reply:
but if you find
```
func addIfNotEmpty(set sets.Set[string], parentNamespace string, ref *common.ResourceRefNamespaced) {
	if !ref.IsEmpty() {
		set.Insert(ref.GetObject(parentNamespace).String())
	}
}
```
and
```
			addIfNotEmpty(result, project.Namespace, &notification.APITokenRef)
			addIfNotEmpty(result, project.Namespace, &notification.DatadogAPIKeyRef)
			addIfNotEmpty(result, project.Namespace, &notification.FlowdockAPITokenRef)
			addIfNotEmpty(result, project.Namespace, &notification.OpsGenieAPIKeyRef)
			addIfNotEmpty(result, project.Namespace, &notification.ServiceKeyRef)
			addIfNotEmpty(result, project.Namespace, &notification.VictorOpsSecretRef)
```

I can change it

Reply:
meh, not a big deal